impress-org/givewp
# Epic: User Telemetery

## Description

This epic introduces non-sensitive information tracker which we will use to improve this plugin.

## Related Addon Epic

https://github.com/impress-org/give-recurring/pull/1035
https://github.com/impress-org/give-fee-recovery/pull/286
https://github.com/impress-org/give-form-field-manager/pull/331

## Affects

This feature does not affect existing functionality.

## Visuals

![image](https://user-images.githubusercontent.com/1784821/109504733-e12c8480-7ac1-11eb-8adc-14894ab332dd.png)
![image](https://user-images.githubusercontent.com/1784821/109795745-54a6d100-7c3d-11eb-9006-21032f77b480.png)

## Pre-review Checklist




## Testing Instructions

- **Test telemetry server onboarding**
    As you can see in the visuals section, an admin will see an admin notice to grant tracking permission. The green color `Opt in` button will onboard the website while other gray links disable notice.
    - `Anonymous Usage Tracking` setting value must set to `enabled` then click on the `Opt in` button.
    - `Anonymous Usage Tracking` setting value must set to `disabled` then click on the gray link (below button).
    
- **Test tracking job**
   A cron job `give_telemetry_send_requests` setup when changes website data, plugin, theme, give wp-settings, and donation form. Also when a new donation created. This job sends data to the telemetry server. You can use any cron manager WordPress plugin to check if the cron job set up correctly or not. This job runs after `24 hours` only if anything changes which we are tracking.
   
- **Tracking events database options**

    A developer can use these options and their value to debug in depth.

| Option Key  | Role |
| ------------- | ------------- |
| give_telemetry_hide_usage_tracking_notice  | This option key contains timestamp or `0` (permanent disable state) which we use to hide telemetry onboarding notice.. |
| give_telemetry_server_access_token  | This option key contains an access token that is required to the community with the telemetry server.  |
| give_telemetry_usage_tracking_last_request  | This option key contains when last request send to telemetry server.  |
| give_telemetry_website_data_checksum  | This option key contains a checksum which we use to detect the change in the website data.  |
| give_telemetry_recently_edited_donation_form  | This option key contains list of donation form which edited recently.  |
| give_telemetry_records  | This option key contains a list of request ids which we will send to the telemetry server.  |

    
**Note:**
1. You can not test this pull request on the local server. Please set up a live site to test it.
2. Recurring donation, fee recovery, and form field manager addon are Telemetry feature compatible. Each addon has an epic pull request as mentioned in the `Related Addon Epic` section. Use this pull request code when test addon compatibility.
